tower-http/src/classify.rs

```diff
@@ -0,0 +1,316 @@
+//! Tools for classifying responses as either success or failure.
+
+use http::{HeaderMap, Request, Response, StatusCode};
+use std::{convert::Infallible, marker::PhantomData};
+
+/// Trait for producing response classifiers from a request.
+///
+/// This is useful if your classifier depends on data from the request. Could for example be the
+/// URI or HTTP method.
+///
+/// This trait is generic over the error type your services produce. This is necessary for
+/// [`ClassifyResponse::classify_error`].
+pub trait MakeClassifier<E> {
+    /// The response classifier produced.
+    type Classifier: ClassifyResponse<
+        E,
+        FailureClass = Self::FailureClass,
+        ClassifyEos = Self::ClassifyEos,
+    >;
+
+    /// The type of failure classifications.
+    type FailureClass;
+
+    /// The type used to classify the response end of stream (EOS).
+    type ClassifyEos: ClassifyEos<E, FailureClass = Self::FailureClass>;
+
+    /// Make a response classifier for this request
+    fn make_classifier<B>(&self, req: &Request<B>) -> Self::Classifier;
+}
+
+/// A [`MakeClassifier`] that works by cloning a classifier.
+///
+/// If you have a [`ClassifyResponse`] that doesn't depend on the request you can use
+/// [`SharedClassifier`] to get a [`MakeClassifier`] for your classifier.
+#[derive(Debug, Clone)]
+pub struct SharedClassifier<C> {
+    classifier: C,
+}
```

Comment:
it seems like this _could_ be a blanket impl for `C: ClassifyResponse<E> + Clone`. I'm not sure whether that's a good idea, since it might conflict with other blanket impls, and this is more explicit, so that's just a thought. We could always add that later if it turns out to be a common use case.

Reply:
Right. I think it would be more useful to keep the blanket impl for closures but I'm not quite sure if they would overlap.